src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -84,84 +55,138 @@ OTARequestorInterface * GetRequestorInstance()
+    fileDesignator              = uri.SubSpan(uri.size() - fileDesignatorLength, fileDesignatorLength);
 
-    // TODO: allow caller to provide their own OTADownloader instance and set BDX parameters
+    return CHIP_NO_ERROR;
+}
+
+void OTARequestor::OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
+{
+    LogQueryImageResponse(response);
 
-    TransferSession::TransferInitData initOptions;
-    initOptions.TransferCtlFlags = chip::bdx::TransferControlFlags::kReceiverDrive;
-    initOptions.MaxBlockSize     = 1024;
-    char testFileDes[9]          = { "test.txt" };
-    initOptions.FileDesLength    = static_cast<uint16_t>(strlen(testFileDes));
-    initOptions.FileDesignator   = reinterpret_cast<uint8_t *>(testFileDes);
+    VerifyOrReturn(context != nullptr, ChipLogError(SoftwareUpdate, "Received QueryImageResponse with invalid context"));
 
-    chip::OperationalDeviceProxy * operationalDeviceProxy = Server::GetInstance().GetOperationalDeviceProxy();
-    if (operationalDeviceProxy != nullptr)
+    OTARequestor * requestorCore = static_cast<OTARequestor *>(context);
+
+    switch (response.status)
     {
-        chip::Messaging::ExchangeManager * exchangeMgr = operationalDeviceProxy->GetExchangeManager();
-        chip::Optional<chip::SessionHandle> session    = operationalDeviceProxy->GetSecureSession();
-        if (exchangeMgr != nullptr && session.HasValue())
-        {
-            exchangeCtx = exchangeMgr->NewContext(session.Value(), &mBdxMessenger);
-        }
+    case EMBER_ZCL_OTA_QUERY_STATUS_UPDATE_AVAILABLE: {
+        VerifyOrReturn(response.imageURI.HasValue(), ChipLogError(SoftwareUpdate, "Update is available but no image URI present"));
 
-        if (exchangeCtx == nullptr)
-        {
-            ChipLogError(BDX, "unable to allocate ec: exchangeMgr=%p sessionExists? %u", exchangeMgr, session.HasValue());
-            return;
-        }
+        // Parse out the provider node ID and file designator from the image URI
+        NodeId nodeId = chip::kUndefinedNodeId;
+        CharSpan fileDesignator;
+        CHIP_ERROR err = requestorCore->ParseBdxUri(response.imageURI.Value(), nodeId, fileDesignator);
+        VerifyOrReturn(err == CHIP_NO_ERROR,
+                       ChipLogError(SoftwareUpdate, "BDX image URI is invalid: %" CHIP_ERROR_FORMAT, err.Format()));
+        requestorCore->mProviderNodeId = nodeId;
+
+        // CSM should already be created for sending QueryImage command so use the same CSM based on the assumption that the
+        // provider node ID that will supply the OTA image is on the same fabric as the sender of the QueryImageResponse
+        requestorCore->ConnectToProvider(kStartBDX);
+        break;
+    }
+    case EMBER_ZCL_OTA_QUERY_STATUS_BUSY:
+        break;
+    case EMBER_ZCL_OTA_QUERY_STATUS_NOT_AVAILABLE:
+        break;
+    // TODO: Add download protocol not supported
```

Comment:
What happens here? Looks like we fall into an unknown state here (i.e. won't retry later if busy, won't tell client if none is available)

Reply:
I'm only adding the case switch here and specifically dealing with the update available case. Handling the other status values is done by: https://github.com/project-chip/connectedhomeip/issues/9524

Reply:
Could you please add TODO comments that reflect the issue #9524, so that other readers of the code can track that and understand it's not ready yet.